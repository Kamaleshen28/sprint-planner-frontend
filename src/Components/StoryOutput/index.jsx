import React, { useContext } from 'react';
import Card from '@mui/material/Card';
import PropTypes from 'prop-types';
import Modal from '@mui/material/Modal';
import CardHeader from '@mui/material/CardHeader';
import Avatar from '@mui/material/Avatar';
import { red } from '@mui/material/colors';
import Divider from '@mui/material/Divider';
import Tile from '../Tile';
import { DataContext } from '../../Contexts/DataContext';
import { CardContent } from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import IconButton from '@mui/material/IconButton';

const Story = ({
  id,
  dependencies,
  startDay,
  endDay,
  developers,
  title,
  assignedDeveloperId,
}) => {
  const { apiResponse, stories } = useContext(DataContext);
  let startDate = new Date(apiResponse.projectStartDate);
  let count = 0;
  while (
    count < startDay ||
    startDate.getDay() === 0 ||
    startDate.getDay() === 6
  ) {
    startDate = new Date(startDate.setDate(startDate.getDate() + 1));
    if (startDate.getDay() != 0 && startDate.getDay() != 6) {
      //Date.getDay() gives weekday starting from 0(Sunday) to 6(Saturday)
      count++;
    }
  }

  const duration = endDay - startDay - 1;
  let endDate = new Date(startDate);

  count = 0;
  while (count < duration) {
    endDate = new Date(endDate.setDate(endDate.getDate() + 1));
    if (endDate.getDay() != 0 && endDate.getDay() != 6) {
      //Date.getDay() gives weekday starting from 0(Sunday) to 6(Saturday)
      count++;
    }
  }

  const [open, setOpen] = React.useState(false);
  const handleOpen = () => {
    setOpen(true);
  };
  const handleClose = () => {
    setOpen(false);
  };

  const style = {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    // width: 400,
    bgcolor: 'background.paper',
    boxShadow: 24,
  };
  return (
    <React.Fragment>
      <Tile leftalign={true} handleOpen={handleOpen}>
        {id}: {title}
      </Tile>
      <Modal
        open={open}
        data-testid="story-modal"
        onClose={handleClose}
        aria-labelledby="child-modal-title"
        aria-describedby="child-modal-description"
      >
        <Card sx={{ ...style, minWidth: 270 }} variant="outlined">
          <CardHeader
            sx={{
              display: 'flex',
              // justifyContent: 'center',
              backgroundColor: 'black',
              // alignItems: 'center',
              color: 'white',
              mx: 'auto',
              border: 'none',

              p: 0,
              pl: 4,
              fontSize: 18,
              fontWeight: 500,

              // paddingBottom: 1.5,
            }}
            avatar={<p>STORY {id}</p>}
            action={
              <p style={{ marginRight: '1rem' }}>
                <IconButton
                  aria-label="settings"
                  onClick={handleClose}
                  sx={{ '&:hover': { bgcolor: '#f9f9f936' } }}
                >
                  <CloseIcon
                    sx={{
                      color: 'white',
                    }}
                  />
                </IconButton>
              </p>
            }
          />
          <Divider variant="middle" />
          <CardContent
            sx={{
              pt: 2,
              px: 4,
              pb: 3,
              width: '370px',
            }}
          >
            {/* <p>
              <b>Story ID:</b> {id}
            </p> */}
            <p>
              <b>Title: </b> {title}
            </p>
            <p>
              <b>Dependencies: </b>
              {dependencies.length !== 0
                ? dependencies.map((dependency, index) => (
                    <li
                      key={index}
                      style={{
                        overflow: 'hidden',
                        whiteSpace: 'nowrap',
                        textOverflow: 'ellipsis',
                        width: '100%',
                      }}
                    >
                      {dependency}:{' '}
                      {stories.find((story) => story.id === dependency).title}
                    </li>
                  ))
                : 'NA'}
            </p>
            <p>
              <b>Start Day: </b> {startDate.toLocaleDateString()}
            </p>
            <p>
              <b>End Day: </b> {endDate.toLocaleDateString()}
            </p>
            <p>
              <b>Developer: </b>
              {developers[0].name}
            </p>
          </CardContent>
        </Card>
      </Modal>
    </React.Fragment>
  );
};
Story.propTypes = {
  title: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  dependencies: PropTypes.array.isRequired,
  startDay: PropTypes.number.isRequired,
  endDay: PropTypes.number.isRequired,
  developers: PropTypes.array.isRequired,
  assignedDeveloperId: PropTypes.number.isRequired,
};
export default Story;
