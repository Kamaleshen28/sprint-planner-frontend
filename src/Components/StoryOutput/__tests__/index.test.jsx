import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import StoryOutput from '..';
import DataProvider from '../../../Contexts/DataContext';
import { useNavigate } from 'react-router-dom';
jest.mock('react-router-dom');

describe('StoryOutput', () => {
  const mockProps = {
    id: 0,
    storyPoints: 4,
    title: 'Authentication',
    dependencies: [1],
    assignedDeveloperId: 0,
    startDay: 2,
    endDay: 4,
    developers: [
      {
        id: 0,
        name: 'nm',
        capacity: 122,
        projectId: 'cc8b748b-8dbb-4f1d-9271-7153d5e0d6cb',
        created_at: '2023-03-16T15:34:57.480Z',
        updated_at: '2023-03-16T15:34:57.480Z',
      },
      {
        id: 1,
        name: 'we',
        capacity: 122,
        projectId: 'cc8b748b-8dbb-4f1d-9271-7153d5e0d6cb',
        created_at: '2023-03-16T15:34:57.480Z',
        updated_at: '2023-03-16T15:34:57.480Z',
      },
    ],
  };
  it('should render the component', () => {
    const navigateFn = jest.fn();
    useNavigate.mockReturnValue(navigateFn);
    const { asFragment } = render(
      <DataProvider
        values={{
          apiResponse: {
            projectStartDate: '2021-03-16T15:34:57.470Z',
          },
        }}
      >
        <StoryOutput {...mockProps} />
      </DataProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('should open Modal to display story details when clicked on specific story', async () => {
    const navigateFn = jest.fn();
    useNavigate.mockReturnValue(navigateFn);
    render(
      <DataProvider
        values={{
          apiResponse: {
            projectStartDate: '2021-03-16T15:34:57.470Z',
          },
        }}
      >
        <StoryOutput {...mockProps} />
      </DataProvider>,
    );
    expect(screen.queryByTestId('story-modal')).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId('modal-button'));
    expect(screen.getByTestId('story-modal')).toBeInTheDocument();
    // fireEvent.click(document.body);
    // await waitFor(() => {
    //   expect(screen.queryByTestId('story-modal')).not.toBeInTheDocument();
    // });
  });
});
