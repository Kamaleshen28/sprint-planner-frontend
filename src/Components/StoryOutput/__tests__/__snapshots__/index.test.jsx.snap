// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryOutput should render the component 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-yczjo3-MuiButtonBase-root-MuiButton-root"
    data-testid="modal-button"
    tabindex="0"
    type="button"
  >
    0: Authentication
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
